### PR TITLE
stabilize combining +bundle and +whole-archive link modifiers

### PR DESCRIPTION
**Pull request by Be-ing**
Per discussion on https://github.com/rust-lang/rust/issues/108081 combining +bundle and +whole-archive already works and can be stabilized independently of other aspects of the packed_bundled_libs feature. There is no risk of regression because this was not previously allowed.

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 1]
```diff
@@ -365,15 +365,9 @@ fn link_rlib<'a>(
     // loaded from the libraries found here and then encode that into the
     // metadata of the rlib we're generating somehow.
     for lib in codegen_results.crate_info.used_libraries.iter() {
-        let NativeLibKind::Static { bundle: None | Some(true), whole_archive } = lib.kind else {
+        let NativeLibKind::Static { bundle: None | Some(true), .. } = lib.kind else {
             continue;
         };
-        if whole_archive == Some(true)
-            && flavor == RlibFlavor::Normal
-            && !codegen_results.crate_info.feature_packed_bundled_libs
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:373]: The `feature_packed_bundled_libs` field can be removed too, it was its only use.

**Comment by petrochenkov** [comment]
The `packed_bundled_libs` feature should also be stabilized (moved from compiler\rustc_feature\src\active.rs to compiler\rustc_feature\src\accepted.rs).

**Comment by Be-ing** [comment]
> The packed_bundled_libs feature should also be stabilized (moved from compiler\rustc_feature\src\active.rs to compiler\rustc_feature\src\accepted.rs).

What about the [other use](https://github.com/rust-lang/rust/blob/3ae0ef79fe4485ad5f0bfc4cd1f27a1b3fd94f60/compiler/rustc_metadata/src/native_libs.rs#L48-L67) of the feature?

**Comment by petrochenkov** [comment]
> > The packed_bundled_libs feature should also be stabilized (moved from compiler\rustc_feature\src\active.rs to compiler\rustc_feature\src\accepted.rs).
> 
> What about the [other use](https://github.com/rust-lang/rust/blob/3ae0ef79fe4485ad5f0bfc4cd1f27a1b3fd94f60/compiler/rustc_metadata/src/native_libs.rs#L48-L67) of the feature?

That's a command line option, not feature.

**Comment by Be-ing** [comment]
Ah, I didn't realize those were turned on in different ways.

**Comment by Be-ing** [comment]
Sorry for letting this sit for a while. @petrochenkov  I have rebased it and addressed review comments.

### Brief summary
Currently, combining +bundle and +whole-archive works only with `#![feature(packed_bundled_libs)]`. This crate feature is independent of the `-Zpacked-bundled-libs` command line option.

This commit stabilizes the `#![feature(packed_bundled_libs)]` crate feature and implicitly enables it only when the +bundle and +whole-archive link modifiers are combined. This allows rlib crates to use the +whole-archive link modifier with native
libraries and have all symbols included in the linked library to be included in downstream staticlib crates that use the rlib as a dependency. Other cases requiring the packed_bundled_libs behavior still require the -Zpacked-bundled-libs command line option, which can be stabilized independently in the future.

Per discussion on https://github.com/rust-lang/rust/issues/108081 there is no risk of regression stabilizing the crate feature in this way because the combination of +bundle,+whole-archive link modifiers was previously not allowed.

### Documentation
I don't think anything needs to be updated for this? From users' perspective it's merely allowing a combination of existing link modifiers that previously failed with an error.

### Tests
[tests/run-make/rlib-format-packed-bundled-libs-3/rust_dep.rs](https://github.com/rust-lang/rust/tree/72e29da3eccd3e4c1fb2c581fa33216db50fcc93/tests/run-make/rlib-format-packed-bundled-libs-3)

Real world example of Rust code requiring this feature: https://github.com/KDAB/cxx-qt/pull/598

### Unresolved questions
The `-Zpacked-bundled-libs` command line option is still unstable.

**Comment by petrochenkov** [comment]
@rfcbot fcp merge
This PR is a promise that the combination `+bundle,+whole-archive` in rlibs will be supported in the future in some way (the specific way in which it's supported stays an implementation detail).

**Comment by apiraino** [comment]
Discussed ion T-compiler meeting [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-08-17/near/385686940), acknowledged that it will need sign-off from T-compiler members.

@rustbot label -I-compiler-nominated

**Pull request merged**